gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopySource.java

```diff
@@ -370,9 +366,25 @@ public Void call() {
               workUnit.setProp(ConfigurationKeys.COPY_EXPECTED_SCHEMA, ((ConfigBasedDataset) this.copyableDataset).getExpectedSchema());
             }
           }
-          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.IS_DATASET_STAGING_DIR_USED,false))) {
-            workUnit.setProp(DATASET_STAGING_DIR_PATH, ((HiveDataset) this.copyableDataset).getProperties().getProperty(DATASET_STAGING_PATH));
+
+          // Ensure that the writer temporary directories are contained within the dataset shard
+          if ((this.copyableDataset instanceof HiveDataset) && (state.getPropAsBoolean(ConfigurationKeys.USE_DATASET_LOCAL_WORK_DIR,false))) {
```

Comment:
It is OK to do it temporarily but using `instanceOf` in `CopySource` definitely seems like a abstraction leaking ( I know it was there already). 

We don't have to address this in this PR, but do you think we should put something like `getDatasetPath` as part of `CopyableDatasetBase` interface?  So that we don't need to explicitly cast them